Readme.md
some bentchmark data can be found in this [repo](https://github.com/sunnycia/Image_compression)

example:

python training.py --training_data_dir='/data/.../CLIC_MOB' --valid_data_dir='' --train_prototxt='prototxt/cwic_train.prototxt' --base_lr=0.00001 --batch=64

history
    1  make -j32
    2   cd ~/pwd/saliency_on_videoset/_Train/
    3  git clone https://github.com/ndrplz/dilation-tensorflow.git
    4  cd dilation-tensorflow/
    5  python main_tf.py 
    6  python
    7  pytho
    8  python
    9  vim ~/.bashrc
   10  source ~/.bashrc
   11  python
   12  pwd
   13  cd /home/sunnycia/pwd/saliency_on_videoset/_Train/Deep-Feature-Flow-master
   14  ls model/
   15  python ./rfcn/demo.py
   16  pip install easydict
   17  python ./rfcn/demo.py
   18  pip install mxnet
   19  ls
   20  python
   21  cd /home/sunnycia/pwd/saliency_on_videoset/_Train/dilation-tensorflow
   22  python
   23  python main_tf.py 
   24  vim ~/.bashrc
   25  source ~/.bashrc
   26  python
   27  pwd
   28  top
   29  cd 
   30  cd pwd/saliency_on_videoset/_Train/dilation-tensorflow/
   31  python
   32  python main_tf.py 
   33  exit()
   34  pip uninstall tensorflow
   35  pip uninstall tensorflow-gpu
   36  cd ~
   37  git clone https://github.com/tensorflow/tensorflow 
   38  unzip tensorflow-master.zip 
   39  cd tensorflow-master/
   40  git checkout
   41  python
   42  cd ~
   43  python
   44  pip install tensorflow
   45  cd pwd/saliency_on_videoset/_Train/dilation-tensorflow/
   46  python main_tf.py 
   47  python
   48  pip install opencv-python
   49  python main_tf.py 
   50  pip uninstall tensorflow
   51  git clone https://github.com/bazelbuild/bazel.git
   52  cd ~
   53  git clone https://github.com/bazelbuild/bazel.git
   54  python
   55  pip install --ignore-installed --upgrade https://storage.googleapis.com/tensorflow/linux/gpu/tensorflow_gpu-1.3.0-cp34-cp34m-linux_x86_64.whl
   56  pip install --ignore-installed https://storage.googleapis.com/tensorflow/linux/gpu/tensorflow_gpu-1.3.0-cp36-cp36m-linux_x86_64.whl
   57  python
   58  locate libcudnn*
   59  ls /usr/local
   60  cd cuda
   61  cd /usr/local/cuda
   62  ls
   63  cd lib
   64  cd lib64/
   65  ls
   66  ls lib cudnn*
   67  ls libcudnn*
   68  ls -l libcudnn*
   69  nvidia-smi
   70  top
   71  cd pwd/saliency_on_videoset/_Train/
   72  ls
   73  git clone https://github.com/sampepose/flownet2-tf.git
   74  ls
   75  ls flow*
   76  unzip flownet2-tf.zip 
   77  cd flownet2-tf/
   78  pip install enum
   79  python
   80  cd checkpoints/
   81  bash download.sh 
   82  vim download.sh 
   83  bash download.sh 
   84  ls
   85  tar -xzvf weights.tar.gz 
   86  cd ..
   87  ls src
   88  python -m src.flownet2.test --input_a data/samples/0img0.ppm --input_b data/samples/0img1.ppm --out ./
   89  python
   90  vim ~/.bahsrc
   91  vim ~/.bashrc
   92  source ~/.bashrc
   93  python -m src.flownet2.test --input_a data/samples/0img0.ppm --input_b data/samples/0img1.ppm --out ./
   94  cd pwd/saliency_on_videoset/_Train/
   95  cd flownet2-tf/
   96  python -m src.flownet2.test --input_a data/samples/0img0.ppm --input_b data/samples/0img1.ppm --out ./
   97  python
   98  vim ~/.bahsrc
   99  vim ~/.bashrc
  100  source ~/.bashrc
  101  python
  102  cd pwd/saliency_on_videoset/_Train/flownet2-tf/
  103  python -m src.flownet2.test --input_a data/samples/0img0.ppm --input_b data/samples/0img1.ppm --out ./
  104  pip install image
  105  python -m src.flownet2.test --input_a data/samples/0img0.ppm --input_b data/samples/0img1.ppm --out ./
  106  make all
  107  nvcc
  108  ls /usr/lib/nvi*
  109  ls /usr/lib
  110  ls nvidia-current
  111  nvcc
  112  vim ~/.bash_profile 
  113  vim ~/.bashrc
  114  source ~/.bashrc
  115  nvcc
  116  cd pwd/saliency_on_videoset/_Train/flownet2-tf/
  117  python -m src.flownet2.test --input_a data/samples/0img0.ppm --input_b data/samples/0img1.ppm --out ./
  118  make clean
  119  make all
  120  make clean
  121  make -j16 all
  122  python -m src.flownet2.test --input_a data/samples/0img0.ppm --input_b data/samples/0img1.ppm --out ./
  123  python -m src.flownet2.test --input_a data/samples/frame88.bmp --input_b data/samples/frame90.bmp --out ./
  124  python -m src.flownet2.test --input_a data/samples/0img0.ppm --input_b data/samples/0img1.ppm --out ./
  125  export CUDA_VISIBLE_DEVICES=6
  126  python -m src.flownet2.test --input_a data/samples/0img0.ppm --input_b data/samples/0img1.ppm --out ./
  127  python -m src.flownet2.test --input_a data/samples/frame88.bmp --input_b data/samples/frame90.bmp --out ./
  128  cd ..
  129  git clone https://github.com/shekkizh/FCN.tensorflow.git
  130  unzip FCN.tensorflow.zip 
  131  cd FCN.tensorflow/
  132  ls
  133  python FCN.py 
  134  cd Data_zoo/MIT_SceneParsing/
  135  unzip ADEChallengeData2016.zip 
  136  cd ../..
  137  python FCN.py 
  138  df -h
  139  cd /
  140  cd /data/
  141  ls
  142  cd sunnycia/
  143  ls
  144  cd YouTube8m/
  145  ls
  146  cd ..
  147  rm -rf YouTube8m/
  148  df -h
  149  pwd
  150  cd ~/pwd/
  151  du -h --max-depth=1 ./
  152  mv saliency_on_videoset/ /data/sunnycia/
  153  df -h
  154  ln -s /data/sunnycia/saliency_on_videoset/ saliency_on_videoset
  155  cd saliency_on_videoset/
  156  pwd
  157  cd pwd/saliency_on_videoset/
  158  ls
  159  cd _Train/
  160  cd FCN.tensorflow/
  161  python FCN.py 
  162  df -h
  163  ls /data
  164  ls /data/zhangpp/
  165  python FCN.py 
  166  cd pwd/saliency_on_videoset/
  167  pwd
  168  cd _Train/FCN.tensorflow/
  169  python FCN.py 
  170  top
  171  nvidia-smi 
  172  export CUDA_VISIBLE_DEVICES=6
  173  python FCN.py 
  174  cd pwd/saliency_on_videoset/_Train/
  175  cd FCN.tensorflow/
  176  python FCN --mode=test
  177  python FCN.py --mode=test
  178  bash ~/setup.sh 
  179  vim ~/setup.sh 
  180  bash ~/setup.sh 
  181  nvidia-smi
  182  top
  183  python FCN.py --mode=test
  184  python FCN.py --mode=visualize
  185  python FCN_inference.py 
  186  export CUDA_VISIBLE_DEVICES=6
  187  nvidia-smi
  188  python FCN_inference.py 
  189  xtermxtermxterm
  190  python FCN_inference.py 
  191  top
  192  nvidia-smi
  193  top
  194  free -h
  195  nvidia-smi
  196  ls
  197  cd pwd/
  198  ln -s /data/sunnycia/pwd/Industry/ Industry

training.py
from Dataset import ImageDataset
import matplotlib
matplotlib.use('TkAgg')
import matplotlib.pyplot as plt
import argparse, cv2, os, glob, sys, time
import cPickle as pkl
import numpy as np
import caffe
from caffe import layers as L
from caffe import params as P
from caffe.proto import caffe_pb2 as P2
from random import shuffle
import google.protobuf.text_format as txtf
from network import cwic_network
from solver import training_solver

from utils import *

caffe.set_mode_gpu()

def get_arguments():
    parser = argparse.ArgumentParser()
    parser.add_argument('--training_data_dir', type=str, required=True, help='training data directory')
    parser.add_argument('--valid_data_dir', type=str, required=True, help='training data directory')
    parser.add_argument('--solver_prototxt', type=str, default='prototxt/solver.prototxt', help='the network prototxt')
    parser.add_argument('--pretrained_model', type=str, default='model/ResNet-50-model.caffemodel', help='Snapshot path.')

    parser.add_argument('--use_snapshot', type=str, default=None, help='Snapshot path.')
    parser.add_argument('--trainingexampleprops', type=float, default=0.8, help="")

    parser.add_argument('--model_name', type=str, default='model', help='Extra model information.')
    parser.add_argument('--extrainfo_str', type=str, default='', help='Extra model information.')
    
    #Network parameter
    parser.add_argument('--batch', type=int, default=16, help='training mini batch')
    parser.add_argument('--feat_channel', type=int, default=64)
    parser.add_argument('--groups', type=int, default=16)
    parser.add_argument('--compression_ratio', type=float, default=1.0)
    parser.add_argument('--pen_weight', type=float, default=0.03)


    #default learning parameter
    parser.add_argument('--train_prototxt', type=str,   default='prototxt/resnet_autoencoder.prototxt', help='Network arch prototxt')
    parser.add_argument('--snapshot_dir',   type=str,   default='training_output', help='Snapshot saving directory.')
    parser.add_argument('--lr_policy',      type=str,   default='fixed')
    parser.add_argument('--stepsize',       type=int,   default=10000)
    parser.add_argument('--base_lr',        type=float, default=0.0001)
    parser.add_argument('--gamma',          type=float, default=0.0001)
    parser.add_argument('--momentum',       type=float, default=0.9)
    
    parser.add_argument('--power',          type=float, default=0.75)
    parser.add_argument('--weight_decay',   type=float, default=5*1e-4)
    
    parser.add_argument('--display',        type=int,   default=100)
    parser.add_argument('--max_iter',       type=int,   default=200000)
    parser.add_argument('--snapshot',       type=int,   default=10000)
    parser.add_argument('--solver_mode',    type=str,   default='GPU')
    parser.add_argument('--solver_type',    type=str,   default='ADAM')
    return parser.parse_args()

print "Parsing arguments..."
args = get_arguments()
mini_batch = args.batch
training_data_dir = args.training_data_dir
train_prototxt_path = args.train_prototxt
solver_prototxt_path = args.solver_prototxt
snapshot_dir= args.snapshot_dir
if not os.path.isdir(snapshot_dir):
    os.makedirs(snapshot_dir)
snapshot_prefix = os.path.join(snapshot_dir, 'snapshot-')
plot_figure_dir = os.path.join(snapshot_dir, 'figure')
if not os.path.isdir(plot_figure_dir):
    os.makedirs(plot_figure_dir)
print "Loss figure will be save to", plot_figure_dir

## Modify Network here
net_proto = cwic_network(batch_size=mini_batch, 
                        output_feat_channel=args.feat_channel, 
                        compression_ratio=args.compression_ratio, 
                        pen_weight=args.pen_weight,
                        groups=args.groups)

with open(train_prototxt_path, 'w') as f:
    f.write(str(net_proto));#exit()
##

## Modify learning parameter here
solver = caffe.proto.caffe_pb2.SolverParameter() 
solver.train_net=train_prototxt_path
solver.snapshot_prefix=snapshot_prefix
solver.base_lr = args.base_lr
# solver.gamma = args.gamma
solver.momentum = args.momentum

# solver.power = args.power
# solver.weight_decay = args.weight_decay

solver.display = args.display
solver.max_iter = args.max_iter
solver.snapshot = args.snapshot
solver.lr_policy = args.lr_policy
solver.solver_mode = P.Solver.SolverMode.DESCRIPTOR.values_by_name[args.solver_mode].number
# solver.solver_type = P.Solver.SolverType.DESCRIPTOR.values_by_name[args.solver_type].number
with open(solver_prototxt_path, 'w') as f:
    f.write(str(solver))
##


## Copy network and  solver to snapshot directory
with open(os.path.join(snapshot_dir, 'network.prototxt'), 'w') as f:
    f.write(str(net_proto))
with open(os.path.join(snapshot_dir, 'solver.prototxt'), 'w') as f:
    f.write(str(solver))
##



pretrained_model_path= args.pretrained_model
snapshot_path = args.use_snapshot
#Check if snapshot exists
if snapshot_path is not None:
    if not os.path.isfile(snapshot_path):
        print snapshot_path, "not exists.Abort"
        exit()
solver=''
if args.solver_type=='SGD':
    solver = caffe.SGDSolver(solver_prototxt_path)
if args.solver_type=='ADADELTA':
    solver = caffe.ADADELTASolver(solver_prototxt_path)
if args.solver_type=='ADAM':
    solver = caffe.AdamSolver(solver_prototxt_path)

if args.use_snapshot == None:
    if pretrained_model_path != '':
        solver.net.copy_from(pretrained_model_path) # untrained.caffemodel
else:
    solver.restore(snapshot_path)

tranining_dataset = ImageDataset(image_dir=training_data_dir)

tart_time = time.time()

max_iter = 50000000
validation_iter = 1000
plot_iter = 500
epoch=10
idx_counter = 0

x=[]
y=[]
z=[] # validation

plt.plot(x, y)
_step=0
if args.use_snapshot != None:
    _step = int(os.path.basename(args.use_snapshot).split('.')[0].split('_')[-1])

print "Start training..."
while _step * mini_batch < max_iter:

    frame_minibatch = tranining_dataset.next_batch(mini_batch)
    # print frame_minibatch.shape;exit()
    solver.net.blobs['data'].data[...] = frame_minibatch
    solver.step(1)

    ##For debug
    # print "Step forward.."
    # solver.net.forward()
    # print "Step backward.."
    # solver.net.backward()


    if _step%plot_iter==0:
        x.append(_step)
        y.append(solver.net.blobs['loss'].data[...].tolist())
        plt.plot(x, y)
        plt.xlabel('Iter')
        plt.ylabel('kld loss')
        plt.savefig(os.path.join(plot_figure_dir, "plot"+str(_step)+".png"))
        plt.clf()

        pkl.dump(x, open(os.path.join(plot_figure_dir, "x.pkl"), 'wb'))
        pkl.dump(y, open(os.path.join(plot_figure_dir, "y.pkl"), 'wb'))


    if _step%validation_iter==0:
        # do validation for validation set, and plot average 
        # metric(cc, sim, auc, kld, nss) performance dictionary
        pass
        # metric_dict = MetricValidation()


    _step+=1
